docs/shadow_paths.md

```diff
@@ -0,0 +1,101 @@
+contains a diagram of the relationship between intended config (`/config`) and
+applied config (`/state`) leaves:
+
+```
+          +---------+
+          |         |    transition intended
+          |intended |    to applied
+          | config  +---------+
+          |         |         |
+          +---------+         |
+              ^               |         config: true
+   +----------|------------------------------------+
+              |               |         config: false
+              |               |
+              |               |
+              |       +-----------------------------+
+              |       |       | operational state   |
+              |       |  +----v----+ +-----------+  |
+              |       |  |         | |           |  |
+              +       |  | applied | |  derived  |  |   operational:true
+            same +------>| config  | |   state   |<-------+
+            leaves    |  |         | |           |  |
+                      |  |         | |           |  |
+                      |  +---------+ +-----------+  |
+                      +-----------------------------+
+```
+
+In this diagram, applied config are "config false", or `/state` leaves that
+mirror intended config leaves, which are "config true", or `/config` leaves. Per
+[design.md](design.md#openconfig-path-compression), one of these are compressed
+out depending on the value of the `-prefer_operational_state` generation flag.
+
+Shadow paths are relevant only for compressed GoStructs, and indicate the
+compressed-out `/config` or `/state` YANG `leaf` nodes.
+
+## Problems with Path Compression and how Shadow Paths Help
+
+Path compression leads to the GoStruct not being able to represent both intended
+config and applied config at the same time. This leads to two problems:
+
+1.  When subscribing to a non-leaf path, some gNMI clients want to silently
+    ignore the compressed-out paths rather than erroring out due to an
+    unrecognized path.
+2.  Some use cases (e.g. [ygnmi](https://github.com/openconfig/ygnmi#queries))
+    use the same compressed GoStruct for representing either the "config view"
+    (intended config+derived state) combination, or the "state view" (applied
+    config+derived state) combination of leaves. We want to enable such a
+    meaning-switch for certain ygot utilities (e.g. marshalling/unmarshalling).
+
+ygot address these issues by,
+
+1.  Always ignoring paths that match a `shadow-path` tag when doing
+    unmarshalling. For example, if a gNMI update for
+    `/interfaces/interface[name="foo"]/config/mtu` is unmarshalled into the
+    `Interface` GoStruct at the beginning of this documentation, then the field
+    will not be populated since it is a shadow path.
+2.  Supporting a `PreferShadowPath` option for some utilities (see section
+    below).
+
+## Preferring Shadow Paths
+
+This term is used to describe utilities preferring the `shadow-path` tag instead
```

Comment:
Which term?

Reply:
`PreferShadowPath` is behavioural option used to describe utilities preferring...